### PR TITLE
[FIX] base, web: should read PropertiesDefinition field from child

### PR DESCRIPTION
**Pull request by ibo-odoo**
**Steps to reproduce:**
Create an article Parent A and set it as private
Create a child article, Child A
Invite a user on Child A so that they can read it
Log in as a user and try to open Child A from either a direct link or a ctrl+k

**Current behavior before PR:**

-It shows an access error regarding the parent article because of the 'Properties
Definition' field.
-The user can't edit the parent but can be able to add property fields from the
sidebar of the child.

**Desired behavior after PR is merged:**

-Without any access error user will be able to read that article, even if it
cannot be displayed in the menu because he can't read its parent.
-If the user can't edit the parent, can not be able to add property fields of the
child.

**Task**-3062064

### CODE REVIEW

File: odoo/fields.py [hunk 1]
```diff
@@ -3452,7 +3452,7 @@ def write(self, records, value):
                 definition.pop('definition_changed', None)
 
             # update the properties definition on the container
-            container = records[self.definition_record]
+            container = records[self.definition_record].sudo()
```
Comment (std-odoo) [inline odoo/fields.py:3455]: I'm ok for the first sudo, but not for this one, here you allow anyone to modify the definition

the definition edition should be disabled if we can't write on the parent (the traceback in the task is only about reading, so this line should not be the issue)

**Review by std-odoo — COMMENTED** [review]
@ibo-odoo Hi :) 

Do we really need the second sudo ?

**Comment by std-odoo** [comment]
can you also add a small test to check we don't loose your fix ?

File: odoo/fields.py (hunk 1 above)
Comment (ibo-odoo) [reply odoo/fields.py]: agree to this.

**Comment by ibo-odoo** [comment]
Hello @std-odoo 
Changes are done. Thank you:)

File: odoo/addons/test_new_api/tests/test_properties.py [hunk 2]
```diff
@@ -1312,6 +1312,17 @@ def _mocked_check_access_rights(operation, raise_exception=True):
             values = self.message_1.read(['attributes'])[0]['attributes'][0]
         self.assertEqual(values['value'], (tag.id, None))
 
+        # a user read a properties with a many2one to a record he has access to
+        # but doesn't have access to its parent
+        self.env.ref('test_new_api.access_discussion').write({
```
Comment (std-odoo) [inline odoo/addons/test_new_api/tests/test_properties.py:1317]: can you use `_mocked_check_access_rights` instead ?

**Review by std-odoo — COMMENTED** [review]
@ibo-odoo Hi :) 

One last comment, otherwise LGTM :)

File: odoo/addons/test_new_api/tests/test_properties.py [hunk 3]
```diff
@@ -1312,6 +1312,19 @@ def _mocked_check_access_rights(operation, raise_exception=True):
             values = self.message_1.read(['attributes'])[0]['attributes'][0]
         self.assertEqual(values['value'], (tag.id, None))
 
+        # a user read a properties with a many2one to a record but doesn't have access to its parent
+        self.env.ref('test_new_api.access_discussion').write({
```
Comment (std-odoo) [inline odoo/addons/test_new_api/tests/test_properties.py:1316]: you mock check_access_rights, you shouldn't need to do this

**Review by std-odoo — COMMENTED** [review]
@ibo-odoo Hi :) 

One more comment about the test

File: addons/web/static/src/views/fields/properties/properties_field.js [hunk 4]
```diff
@@ -368,11 +368,27 @@ export class PropertiesField extends Component {
...
-        this.state.canChangeDefinition = await this.orm.call(
+        const hasAccessRights = await this.orm.call(
             definitionRecordModel,
             "check_access_rights",
             ["write", false],
         );
+
+        let checkAccessRule; // undefined if the operation is allowed
+        try {
+            checkAccessRule = await this.orm.call(
```
Comment (std-odoo) [inline addons/web/static/src/views/fields/properties/properties_field.js:379]: no need to do this call is hasAccessRights is false

**Review by std-odoo — COMMENTED** [review]
@ibo-odoo Bloup

File: addons/web/static/src/views/fields/properties/properties_field.js [hunk 5]
```diff
@@ -368,11 +368,27 @@ export class PropertiesField extends Component {
...
+                [definitionRecordId],
+                {
+                    operation: "write",
+                }
+            );
+        } catch (_e) {
+            checkAccessRule = true;
+        }
+
+        this.state.canChangeDefinition = hasAccessRights && !checkAccessRule;
```
[the remark this answers is not in the document]
Comment (ibo-odoo) [reply addons/web/static/src/views/fields/properties/properties_field.js:391]: True that but 'check_access_rule' will return NONE if the operation is allowed so checkAccessRule will be undefined and if I will pass checkAccessRule then `canChangeDefinition` become false which is not correct i think i should change the name instead. What's say?

[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/web/static/src/views/fields/properties/properties_field.js:391]: or just set it to true by default

**Comment by tli-asl** [comment]
Hello there 👋 ,

Any update regarding this issue ? It has also a important impact regarding task/project management (In that context, coming from v15.0, it's more a regression than a side effect on a new module #113440 ).

Have a nice Friday,

cc: @std-odoo

**Comment by xavierbol** [comment]
Hello 👋🏼 ,

Is it possible to have a review for this fix asap? 🙂 

This issue is a bit annoying for the services app. 😕 😢 

Thanks in advance! 

Thanks @ibo-odoo for the fix! 🙂 

Have a nice Friday!

**Comment by tli-asl** [comment]
:D :D :D

https://tenor.com/view/caveira-skeleton-ive-been-waiting-gif-13523984

**Comment by awa-odoo** [comment]
Hello @tli-asl @xavierbol 

TL;DR: It's currently blocked for performance reasons.

I made an alternative proposal, you can check the chatter of task-3062064 for more details (well, if you have access to that :upside_down_face: ).

Cheers!

**Review by std-odoo — COMMENTED** [review]
@ibo-odoo Hi :) 

Some comments / suggestions

Cheers

File: addons/web/static/src/views/fields/properties/properties_field.js [hunk 6]
```diff
@@ -24,6 +24,7 @@ export class PropertiesField extends Component {
         this.dialogService = useService("dialog");
         this.popover = usePopover();
         this.propertiesRef = useRef("properties");
+        this.propertyParentAccessByRecord = {};
```
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/web/static/src/views/fields/properties/properties_field.js:27]: Actually the field is not instantiated once per record.

For example when clicking on a list item and opening its form view, the field will be instantiated.
But then if you click on the "left/right" arrow to switch from one record to another then the form view just "loads" the record, it does not re-instantiate the field(s).

Same thing for knowledge articles, this technique allows a very fast switch between records.

That's why I suggested to IBO to use a dict-like storage that would only check once per record.

[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/web/static/src/views/fields/properties/properties_field.js:27]: ah yes indeed

Isn't this cache system a bit overkill ? (and not that's it's not 100% correct either, you technically can have access rules on the parent based on some field on the child, you can change them and boum you can write on the parent but the cache is not updated, in prastise it won't happen I think)

[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/web/static/src/views/fields/properties/properties_field.js:27]: The idea was to not make a RPC every time you click on a button to modify a add//modify/remove a property.
But I guess in practice people will only try once and see that they don't have access, so I'd be ok with simplifying.

[the remark this answers is not in the document]
Comment (ibo-odoo) [reply addons/web/static/src/views/fields/properties/properties_field.js:27]: @awa-odoo @std-odoo  What is the final decision for `propertyParentAccessByRecord`?

[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/web/static/src/views/fields/properties/properties_field.js:27]: @ibo-odoo To simplify and not use cache I think :) (the simpler the better)

File: addons/web/static/src/views/fields/properties/properties_field.js [hunk 7]
```diff
@@ -296,6 +302,32 @@ export class PropertiesField extends Component {
...
+                [definitionRecordId],
+                {
+                    operation: "write",
+                }
+            );
+        } catch (_e) {
+            this.notification.add(_lt("You need to be able to edit parent first to configure property fields"), {
+                type: "warning",
+            });
+            checkAccessRule = false;
```
Comment (std-odoo) [inline addons/web/static/src/views/fields/properties/properties_field.js:326]: you can directly return false, no need to create the variable `checkAccessRule`

File: addons/web/static/src/views/fields/properties/properties_field.js [hunk 8]
```diff
@@ -296,6 +302,32 @@ export class PropertiesField extends Component {
...
+        }
```
Comment (std-odoo) [inline addons/web/static/src/views/fields/properties/properties_field.js:327]: shouldn't we also change `this.state.canChangeDefinition` ? (so the pencil / create button become invisible)

File: addons/web/static/src/views/fields/properties/properties_field.js [hunk 9]
```diff
@@ -462,6 +494,7 @@ export class PropertiesField extends Component {
             {
                 readonly: this.props.readonly || !this.state.canChangeDefinition,
                 canChangeDefinition: this.state.canChangeDefinition,
+                canModifyTags: () => this.canModifyPropertiesDefinition(),
```
Comment (std-odoo) [inline addons/web/static/src/views/fields/properties/properties_field.js:497]: can you keep the name `canModifyPropertiesDefinition` ? (so no `canModifyTags`, easier to grep and follow the path)

also I would call it `checkDefinitionWriteAccess` or something like that, to not conflict to much with `canChangeDefinition`

**Review by std-odoo — COMMENTED** [review]
@ibo-odoo Hi :) 

Some comments on the latest changes

Cheers

File: addons/web/static/src/views/fields/properties/properties_field.js (hunk 8 above)
Comment (ibo-odoo) [reply addons/web/static/src/views/fields/properties/properties_field.js:327]: I think we don't need to change `this.state.canChangeDefinition` because AWA doesn't want to make pencil/create buttons invisible for the users that don't have access to the parent. That is why we do all this and add a notification while the user clicks on the pencil/create button. If I am wrong then please correct me @awa-odoo

Comment (std-odoo) [reply addons/web/static/src/views/fields/properties/properties_field.js:326]: The state will be only for the current user no ? Or you mean for all parents ? (if yes I understand we don't want to do it, a bit annoying as we will have to do something on `onWillUpdateProps`)

Comment (awa-odoo) [reply addons/web/static/src/views/fields/properties/properties_field.js:326]: I think it's strange to have the buttons present and then having them disappear when used once.

We can probably assume that people are not going to click more than once after they see the error message.

**Comment by awa-odoo** [comment]
@robodoo retry

**Comment by awa-odoo** [comment]
@robodoo r+

note: staging conflicted because of recently merged changes on the properties field (type many2one) "wrong model" fix.
